src/panoptes/pocs/focuser/birger.py

```diff
@@ -246,39 +250,62 @@ def _send_command(self, command, response_length=None):
+            self._serial_port.reset_input_buffer()
 
-        # Adaptor should then send 'OK', even if there was an error.
-        ok = self._serial_io.readline().rstrip()
-        assert ok == 'OK'
+            # Send the command
+            self._serial_io.write(command + '\r')
 
-        # Depending on which command was sent there may or may not be any further
-        # response.
-        response = []
+            # In verbose mode adaptor will first echo the command
+            echo = self._serial_io.readline().rstrip()
 
-        if response_length == 0:
-            # Not expecting any further response. Should check the buffer anyway in case an error
-            # message has been sent.
-            if self._serial_port.in_waiting:
-                response.append(self._serial_io.readline())
+            # Adaptor should then send 'OK', even if there was an error.
+            ok = self._serial_io.readline().rstrip()
 
-        elif response_length > 0:
-            # Expecting some number of lines of response. Attempt to read that many lines.
-            for i in range(response_length):
-                response.append(self._serial_io.readline())
+            if echo == command and ok == 'OK':
+                if response_length == 0:
+                    # Not expecting any further response. Should check the buffer anyway in case an error
+                    # message has been sent.
+                    if self._serial_port.in_waiting:
+                        response.append(self._serial_io.readline())
 
-        else:
-            # Don't know what to expect. Call readlines() to get whatever is there.
-            response.append(self._serial_io.readlines())
+                elif response_length > 0:
+                    # Expecting some number of lines of response. Attempt to read that many lines.
+                    for i in range(response_length):
+                        response.append(self._serial_io.readline())
+
+                else:
+                    # Don't know what to expect. Call readlines() to get whatever is there.
+                    response.append(self._serial_io.readlines())
+
+                try:
+                    assert (echo, ok) == (command, 'OK')
+                    success = True
+                except AssertionError as err:
+                    self.logger.error(err)
+                    warn(err)
+
+                self.logger.trace(f"Got correct response after {i + 1} attempts")
+                break
+
+            if echo != command:
+                self.logger.warning(f"echo != command: {echo!r} != {command!r}. Retrying command.")
+
+            if ok != 'OK':
+                self.logger.warning(f"ok != 'OK': {ok!r} != 'OK'. Retrying command.")
+
+        try:
+            assert(success)
+        except AssertionError:
+            raise error.PanError(f'The command {command!r} was not properly read by {self}')
```

Comment:
With all the edits this has ended up a bit messy. Suggestion to clean it up:

```
        response = []
        success = False

        for i in range(self._max_command_retries):
            self._serial_port.reset_input_buffer()
            self._serial_io.write(command + '\r')

            echo = self._serial_io.readline().rstrip()
            if echo != command:
                self.logger.warning(f"echo != command: {echo!r} != {command!r}. Retrying command.")
                continue

            ok = self._serial_io.readline().rstrip()
            if ok != 'OK':
                self.logger.warning(f"ok != 'OK': {ok!r} != 'OK'. Retrying command.")
                continue

            if response_length == 0:
                # Not expecting any further response. Should check the buffer anyway in case an error
                # message has been sent.
                if self._serial_port.in_waiting:
                    response.append(self._serial_io.readline())
            elif response_length > 0:
                # Expecting some number of lines of response. Attempt to read that many lines.
                for i in range(response_length):
                     response.append(self._serial_io.readline())
            else:
                # Don't know what to expect. Call readlines() to get whatever is there.
                response.append(self._serial_io.readlines())

            success = True
            break

        if not success:
            raise error.PanError(f'Failed command {command!r} on {self}')

```